### PR TITLE
[NX] add Spec for Nx CLI

### PR DESCRIPTION
**Pull request by ronderksen**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Feature

**What is the current behavior? (You can also link to an open issue here)**
Fig currently has no spec for use with the Nx CLI.

**What is the new behavior (if this is a feature change)?**
Introduces autocomplete support for the Nx CLI. The CLI has quite a few commands, which can be hard to remember. 

**Additional info:**
There's still a few commands missing. I will add them in the next few days.

### CODE REVIEW

File: dev/nx.ts [hunk 1]
```diff
@@ -0,0 +1,492 @@
+const nxGenerators: Record<string, Fig.Generator> = {
```
Comment (cstrnt) [inline dev/nx.ts:1]: using Record<string, Fig.Generator> causes TypeScript to also allow nxGenerators.nonExistingProperty. Even though it looks hacky, I would prefer not typing the object itself but rather each value as` Fig.Generator`

File: dev/nx.ts [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+    },
+    postProcess: function (out) {
+      return out
+        .split("\n")
+        .map((line) => line.split(" ").pop())
+        .map((suggestion) => ({
+          name: suggestion,
+          type: "option",
+        }));
+    },
```
Comment (cstrnt) [inline dev/nx.ts:84]: This is basically the same as the generator for `localSchematics`.  I would prefer referencing the cache and postProcess stuff from the `localSchematics` generator to reduce code doublication

File: dev/nx.ts [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+  },
+  installedPlugins: {
+    script: () => {
+      return "nx list";
+    },
```
Comment (cstrnt) [inline dev/nx.ts:89]:
```suggestion
    script: "nx list" ,
```

File: dev/nx.ts [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+    cache: {
+      ttl: 60 * 60 * 24,
+    },
```
Comment (cstrnt) [inline dev/nx.ts:92]: The cache property is re-used. I would prefer to create a top-level const with the type `Fig.Cache`

File: dev/nx.ts [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+      if (argument.indexOf(":") > -1) {
+        return `nx list ${argument.split(":")[0]}`;
+      } else {
+        return "nx list";
+      }
+    },
+    trigger: ":",
+    cache: {
+      ttl: 60 * 60 * 24,
+    },
```
Comment (cstrnt) [inline dev/nx.ts:119]: see above

**Review by cstrnt — CHANGES_REQUESTED** [review]
What a huge PR! Thank you very much 😊

File: dev/nx.ts [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+      .map((suggestion) => ({
+        name: suggestion,
+        type: "option",
+      }));
+  } catch (err) {
+    return [];
+  }
+};
+
+const processGenerators: PostProcessFn = function (out) {
```
Comment (QuiiBz) [inline dev/nx.ts:31]: Shouldn't this be a `() => {...}` instead of a `function`, since you mostly used them?

File: dev/nx.ts [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+  },
+  localGenerators: {
+    script: "ls tools/generators",
+    cache: oneDayCache,
+    postProcess: processGenerators,
+  },
+  installedPlugins: {
+    script: "nx list",
+    cache: oneDayCache,
+    postProcess: function (out) {
```
Comment (QuiiBz) [inline dev/nx.ts:79]: Same here, I would prefer not to mix the two and instead choose one.

**Review by QuiiBz — APPROVED** [review]

File: dev/nx.ts [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+    }));
+};
+
+const oneDayCache: Fig.Cache = {
+  ttl: 60 * 60 * 24,
+};
+
+const nxGenerators: NxGenerators = {
+  apps: {
+    script: "cat workspace.json",
```
Comment (mattschrage) [inline dev/nx.ts:47]: I am not super familiar with the `nx` tool - is it possible to run `nx` commands from child directories of the root folder that contains the `workspace.json` file?

If this is the case, you might be able to adapt the same script that the `npm` CLI uses to find the `package.json` file:

```
until [[ -f package.json ]] || [[ $PWD = '/' ]]; do cd ..; done; cat package.json
```

**Review by mattschrage — COMMENTED** [review]
This looks great! Thanks for contributing @ronderksen :)

I had one small question, but I think this is about ready to merge in.

File: dev/nx.ts (hunk 8 above)
Comment (ronderksen) [reply dev/nx.ts:47]: No, it only works from the root folder.

**Comment by mattschrage** [comment]
@ronderksen sorry for the delay on this! Everything looks good to me. Going to merge now!

**Pull request merged**